User-guide: Add documentation for CAPM3 Remediation feature (#248)
metal3-io/metal3-docs: docs/user-guide/src/capm3/remediaton.md

# Remediation Controller and MachineHealthCheck

> The Cluster API includes [MachineHealthCheck(MHC)](https://cluster-api.sigs.k8s.io/tasks/healthcheck.html) resource that implements automated health checking feature.  It is possible to plug-in Metal3 specific remediation strategies to remediate unhealthy nodes. In this case, Cluster API MHC finds unhealthy nodes while CAPM3 Remediation Controller remediates those unhealthy nodes.

Author: Done

> The Cluster API includes the [MachineHealthCheck(MHC)](https://cluster-api.sigs.k8s.io/tasks/healthcheck.html) feature that implements an automated health checking. Basically, it deletes unhealthy machine and replace with new healthy one. This approach can be challenging with cloud providers that are using hardware based clusters. To overcome this situation, CAPI MachineHealthCheck feature allows provider specific external remediation. So it is also possible to plug-in Metal3 specific remediation strategies to remediate unhealthy nodes. In this case, the Cluster API MHC finds unhealthy nodes while the CAPM3 Remediation Controller remediates those unhealthy nodes.

Reviewer: ```suggestion
The Cluster API includes the [remediation](https://cluster-api.sigs.k8s.io/tasks/healthcheck.html) feature that implements an automated health checking of k8s nodes. It deletes unhealthy Machine and replaces with a healthy one. This approach can be challenging with cloud providers that are using hardware based clusters. To overcome this situation, CAPI remediation feature was extended to plug-in provider specific external remediation. It is also possible to plug-in Metal3 specific remediation strategies to remediate unhealthy nodes. In this case, the Cluster API MHC finds unhealthy nodes while the CAPM3 Remediation Controller remediates those unhealthy nodes.
```

> This approach can be challenging with cloud providers ...

I think it is worth of mentioning why.

> ## CAPI MachineHealthCheck
> 

Reviewer: ```suggestion
## CAPI Remediation

```

> CAPM3 Remediation Controller(RC) reconciles Metal3Remediation objects created by CAPI MachineHealthCheck. The RC locates a Machine with the same name as the Metal3Remediation CR and uses BMO and CAPM3 APIs to remediate associated unhealthy baremetal nodes. Remediation controller supports reboot strategy specified in Metal3Remediation CRD and uses the same object to store state of the current remediation cycle.

Author: It should be nodes. As if the node is unhealthy and doesn't join the cluster then CAPI consider the machine as unhealthy.

> CAPM3 Remediation Controller(RC) reconciles Metal3Remediation objects created by CAPI MachineHealthCheck. The RC locates a Machine with the same name as the Metal3Remediation CR and uses BMO and CAPM3 APIs to remediate associated unhealthy baremetal nodes. Remediation controller supports reboot strategy specified in Metal3Remediation CRD and uses the same object to store state of the current remediation cycle.

Author: Done

## CAPI MachineHealthCheck

> A MachineHealthCheck is a Cluster API resource, which allows users to define conditions under which Machines within a Cluster should be considered unhealthy. Users can also specify a timeout for each of the conditions that they define to check on the Machine’s Node. If any of these conditions are met for the duration of the timeout, the Machine will be remediated. CAPM3 will use MHC to create remediation requests based on Metal3RemediationTemplate and Metal3Remediation CRDs to plug-in remediation solution. For more info, please read the [CAPI MHC)](https://cluster-api.sigs.k8s.io/tasks/healthcheck.html) link.

Author: Machine health check is a CAPI feature and CAPM3 is using the CAPI MHC feature and add the Metal3Remediation object by using the the CR.

> External remediation provides remediation solution other than deleting unhealthy Machine and creating new one. Environments consisting of hardware based clusters are slower to (re)provision unhealthy Machines. So there is a growing need for a remediation flow that includes external remediation which can significantly reduce the remediation process time. Normally the conditions based remediation doesn’t offer any other remediation than deleting an unhealthy Machine and replacing it with a new one. Other environments and vendors can also have specific remediation requirements, so there is a need to provide a generic mechanism for implementing custom remediation logic. External remediation integrates with CAPI MHC and support remediation based on power cycling the underlying hardware. It supports use of BMO reboot API and CAPM3 unhealthy annotation as part of automated remediation cycles. It is a generic mechanism for supporting externally provided custom remediation strategies. If no value for externalRemediationTemplate is defined for the MachineHealthCheck CR, the condition-based flow is continued. For more info: [External Remediation proposal](https://github.com/kubernetes-sigs/cluster-api/pull/3190/files)

Reviewer: ```suggestion
External remediation provides remediation solution other than deleting unhealthy Machine and creating new one. Environments consisting of hardware based clusters are slower to (re)provision unhealthy Machines. So there is a growing need for a remediation flow that includes external remediation which can significantly reduce the remediation process time. Normally the conditions based remediation doesn’t offer any other remediation than deleting an unhealthy Machine and replacing it with a healthy one. Other environments and vendors can also have specific remediation requirements, so there is a need to provide a generic mechanism for implementing custom remediation logic. External remediation integrates with CAPI MHC and support remediation based on power cycling the underlying hardware. It supports use of BMO reboot API and CAPM3 unhealthy annotation as part of automated remediation cycles. It is a generic mechanism for supporting externally provided custom remediation strategies. If no value for externalRemediationTemplate is defined for the MachineHealthCheck CR, the condition-based flow is continued. For more info: [External Remediation proposal](https://github.com/kubernetes-sigs/cluster-api/pull/3190/files)
```

> ## Metal3 Remediation feature
> 
> The CAPM3 remediation Controller reconciles Metal3Remediation objects created by CAPI MachineHealthCheck. It locates a Machine with the same name as the Metal3Remediation object and uses BMO and CAPM3 APIs to remediate associated unhealthy nodes. The remediation controller supports a reboot strategy specified in the Metal3Remediation CRD and uses the same object to store state of the current remediation cycle.

Reviewer: ```suggestion
## Metal3 Remediation

The CAPM3 remediation controller reconciles Metal3Remediation objects created by CAPI MachineHealthCheck. It locates a Machine with the same name as the Metal3Remediation object and uses BMO and CAPM3 APIs to remediate associated unhealthy nodes. The remediation controller supports a reboot strategy specified in the Metal3Remediation CRD and uses the same object to store state of the current remediation cycle.
```

> ### Enable MachineHealthCheck and remediation for worker nodes
> 

Reviewer: ```suggestion
### Enable remediation for worker nodes

```

  # the Node it created has not yet registered with the cluster. This can be useful when
  # Nodes take a long time to start up or when you only want condition based checks for
  # Machine health.
  nodeStartupTimeout: 0m
  # selector is used to determine which Machines should be health checked
  selector:
    matchLabels:
      nodepool: nodepool-0
  # Conditions to check on Nodes for matched Machines, if any condition is matched for the duration of its timeout, the Machine is considered unhealthy
  unhealthyConditions:
  - type: Ready
    status: Unknown
    timeout: 300s
  - type: Ready
    status: "False"
    timeout: 300s
  remediationTemplate: # added infrastructure reference
    kind: Metal3RemediationTemplate
    apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
    name: worker-remediation-request

```

Metal3RemediationTemplate for worker nodes:

```yaml

apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: Metal3RemediationTemplate
metadata:
    name: worker-remediation-request
    namespace: metal3
spec:
  template:
    spec:
      strategy:
        type: "Reboot"
        retryLimit: 2
        timeout: 300s

```

> ### Enable MachineHealthCheck and remediation for control plane nodes
> 

Reviewer: ```suggestion
### Enable remediation for control plane nodes

```

# Remediation Controller and MachineHealthCheck

> The ```Cluster API``` includes [MachineHealthCheck(MHC)](https://cluster-api.sigs.k8s.io/tasks/healthcheck.html) resource that implements automated health checking feature.  It is possible to plug-in Metal3 specific remediation strategies to remediate an unhealthy nodes while depending on ```Cluster API``` MHC resource to determine those nodes as unhealthy by using ```CAPM3 Remediation Controller```.

Reviewer 2: ```suggestion
The ```Cluster API``` includes [MachineHealthCheck(MHC)](https://cluster-api.sigs.k8s.io/tasks/healthcheck.html) resource that implements automated health checking feature.  It is possible to plug-in Metal3 specific remediation strategies to remediate unhealthy nodes while depending on ```Cluster API``` MHC resource to determine those nodes as unhealthy by using ```CAPM3 Remediation Controller```.
```

Author: done

> The ```Cluster API``` includes [MachineHealthCheck(MHC)](https://cluster-api.sigs.k8s.io/tasks/healthcheck.html) resource that implements automated health checking feature.  It is possible to plug-in Metal3 specific remediation strategies to remediate an unhealthy nodes while depending on ```Cluster API``` MHC resource to determine those nodes as unhealthy by using ```CAPM3 Remediation Controller```.

Author: Thanks for the review, done.

## Remediation Controller

> CAPM3 ```Remediation Controller(RC)``` reconciles ```Metal3Remediation``` objects created by CAPI MachineHealthCheck. The RC locates a Machine with the same name as the Metal3Remediation CR and uses BMO and CAPM3 APIs to remediate associated unhealthy baremetal nodes. Remediation controller supports ```reboot strategy``` specified in Metal3Remediation CRD and uses the same object to store state of the current remediation cycle.

Author: done

## CAPI MachineHealthCheck

A MachineHealthCheck is a Cluster API resource, which allows users to define conditions under which Machines within a Cluster should be considered unhealthy. Users can also specify a timeout for each of the conditions that they define to check on the Machine’s Node. If any of these conditions are met for the duration of the timeout, the Machine will be remediated. CAPM3 will use MHC to create remediation requests based on ```Metal3RemediationTemplate``` and ```Metal3Remediation``` CRDs to plug-in remediation solution.

### Remediation workflow

* CAPM3 ```Remediation Controller(RC)``` detects the presence of ```Metal3Remediation``` CR.
* Remediation strategy can be defined under ```.spec.strategy.type``` in Metal3Remediation, RC uses BMO APIs to get hosts back into a healthy or manageable state.
* RC uses ```.status.phase``` to save the states of the remediation. Available states are ```running```, ```waiting```, ```deleting machine```.

> * After RC have finished its remediation, it will wait for the Metal3Remediation CR to be removed. MHC will noticed the Node becomes healthy and deletes the instantiated MachineRemediation CR using CAPI MachineHealthCheck controller.

Reviewer 2: ```suggestion
* After RC has finished its remediation, it will wait for the Metal3Remediation CR to be removed. MHC will notice the Node to become healthy and delete the instantiated MachineRemediation CR using CAPI MachineHealthCheck controller.
```

Author: This part is removed

> * ```.spec.strategy.retryLimit``` and  ```.spec.strategy.timeout``` defined in Metal3Remediation are used to set limit for reboot retries and time between retries. If RCs last ```.spec.strategy.timeout``` for Node to become healthy expires, it sets ```capi.MachineOwnerRemediatedCondition``` to False on Machine object to start deletion of the unhealthy Machine and the corresponding Metal3Remediation. Also it annotates BareMetalHost with ```capi.metal3.io/unhealthyannotation```.

Reviewer 2: ```suggestion
* ```.spec.strategy.retryLimit``` and  ```.spec.strategy.timeout``` defined in Metal3Remediation are used to set limit for reboot retries and time between retries. If RCs last ```.spec.strategy.timeout``` for Node to become healthy expires, it sets ```capi.MachineOwnerRemediatedCondition``` to False on Machine object to start deletion of the unhealthy Machine and the corresponding Metal3Remediation. Also, it annotates BareMetalHost with ```capi.metal3.io/unhealthyannotation```.
```

Author: Removed

        retryLimit: 2
        timeout: 300s

```

### Configuration for controlplane nodes

User can apply the following examples for creating a MachineHealthCheck and Metal3Remediation for controlplane nodes:

```yaml

apiVersion: cluster.x-k8s.io/v1beta1
kind: MachineHealthCheck
metadata:
  name: controlplane-healthcheck
  namespace: metal3
spec:
  clusterName: test1
  maxUnhealthy: 100%
  nodeStartupTimeout: 0m
  selector:
    matchLabels:
      cluster.x-k8s.io/control-plane: ""
  unhealthyConditions:
    - type: Ready
      status: Unknown
      timeout: 300s
    - type: Ready
      status: "False"
      timeout: 300s
  remediationTemplate: # added infrastructure reference
    kind: Metal3RemediationTemplate
    apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
    name: controlplane-remediation-request

```

```Metal3RemediationTemplate``` for controlplane nodes:

```yaml

apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: Metal3RemediationTemplate
metadata:
    name: controlplane-remediation-request
    namespace: metal3
spec:
  template:
    spec:
      strategy:
        type: "Reboot"
        retryLimit: 1
        timeout: 300s

```

## Limitations and Caveats

> * Machines owned by a MachineSet or a KubeadmControlPlane can be remediated by a MachineHealthCheck

Reviewer 2: Is this a limitation?

Author: It can be both limitation and  caveats(warnings). As machine owned by machineset or kcp can be remediated.
